core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -0,0 +1,476 @@
+        Collections.unmodifiableList(new ArrayList<>(servers));
+
+    if (addressIndex == null) {
+      addressIndex = new Index(newImmutableAddressGroups);
+    } else if (currentState == READY) {
+      // If a ready subchannel exists in new address list,
+      // keep this connection and don't create new subchannels
+      SocketAddress previousAddress = addressIndex.getCurrentAddress();
+      addressIndex.updateGroups(newImmutableAddressGroups);
+      if (addressIndex.seekTo(previousAddress)) {
+        return true;
+      }
+      addressIndex.reset();
+    } else {
+      addressIndex.updateGroups(newImmutableAddressGroups);
+    }
+
+    // Create subchannels for all new addresses, preserving existing connections
+    Set<SocketAddress> oldAddrs = new HashSet<>(subchannels.keySet());
+    Set<SocketAddress> newAddrs = new HashSet<>();
+    for (EquivalentAddressGroup endpoint : newImmutableAddressGroups) {
+      for (SocketAddress addr : endpoint.getAddresses()) {
+        newAddrs.add(addr);
+        if (!subchannels.containsKey(addr)) {
+          createNewSubchannel(addr);
+        }
+      }
+    }
+
+    // remove old subchannels that were not in new address list
+    for (SocketAddress oldAddr : oldAddrs) {
+      if (!newAddrs.contains(oldAddr)) {
+        subchannels.get(oldAddr).getSubchannel().shutdown();
+        subchannels.remove(oldAddr);
+      }
+    }
+
+    if (oldAddrs.size() == 0 || currentState == CONNECTING || currentState == READY) {
+      // start connection attempt at first address
+      updateBalancingState(CONNECTING, new Picker(PickResult.withNoResult()));
+      requestConnection();
+
+    } else if (currentState == IDLE) {
+      // start connection attempt at first address when requested
+      SubchannelPicker picker = new RequestConnectionPicker(this);
+      updateBalancingState(IDLE, picker);
+
+    } else if (currentState == TRANSIENT_FAILURE) {
+      // start connection attempt at first address
+      requestConnection();
+    }
+
+    return true;
+  }
+
+  @Override
+  public void handleNameResolutionError(Status error) {
+    for (SubchannelData subchannelData : subchannels.values()) {
+      subchannelData.getSubchannel().shutdown();
+    }
+    subchannels.clear();
+    // NB(lukaszx0) Whether we should propagate the error unconditionally is arguable. It's fine
+    // for time being.
+    updateBalancingState(TRANSIENT_FAILURE, new Picker(PickResult.withError(error)));
+  }
+
+  void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateInfo) {
+    ConnectivityState newState = stateInfo.getState();
+    // Shutdown channels/previously relevant subchannels can still callback with state updates.
+    // To prevent pickers from returning these obselete subchannels, this logic
+    // is included to check if the current list of active subchannels includes this subchannel.
+    if (!subchannels.containsKey(getAddress(subchannel))) {
```

Comment:
Fixed it to this:

`    if (!subchannels.containsKey(getAddress(subchannel))
        || subchannels.get(getAddress(subchannel)).getSubchannel() != subchannel) {
      return;
    }
`

As an fyi, I needed to keep the containsKey() to avoid nullpointers when calling getSubchannel()